### PR TITLE
mgr/cephadm: reconfig prometheus when service-discovery port changes

### PR DESCRIPTION
**Pull request by rkachach**
Some refactoring + Adding logic to restart service-discovery http server and redeploy Prometheus when service discovery port is changed.

fixes: https://tracker.ceph.com/issues/57366

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test make check

**Review by epuertat — APPROVED** [review]
LGTM @rkachach ! Just a few suggestions over there.

File: src/pybind/mgr/cephadm/service_discovery.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/service_discovery.py:14]: It sounds as good improvement to avoid importing explicit types but I think it w'd break the style (in comparison with other cephadm files). So in case we go and introduce something like that I'd do it for all the files at once to keep the style similar across the code.

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py]: Moved

File: src/pybind/mgr/cephadm/service_discovery.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/service_discovery.py]: I love the idea, code updated.

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py]: Updated the code accordingly

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py:24]: @epuertat I'm not sure if this variable is actually shared between modules. Isn't each module using its own sub-interpreter?

Comment (epuertat) [reply src/pybind/mgr/cephadm/http_server.py:24]: I didn't mean to share the variable itself, but the code to expose it as a module option (similar to the log level module option, which is set in the mgr_module file, but can be tweaked per module), but anyway not a critical thing.

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py:74]: @epuertat during my testing I didn't see any suspicious behavior. At the begining I ran in some issues because when trying to restart the server (and bind the same port) I was getting port in use error because the actually the server didn't free the port. I add to add this callback in order for force calling the `unsbuscribe()` when stopping the server:

```
    def stop(self) -> None:
        # we must call unsubscribe before stopping the server,
        # otherwise the port is not released and we will get
        # an exception when trying to restart it
        self.unsubscribe()
        super().stop()
```
basides, in a previour PR I also found that setting the cherrypy server to None as in the following code helped to free stuff and re-create the server:

```
    def restart(self) -> None:
        cherrypy.engine.stop()
        cherrypy.server.httpserver = None
        self.configure()
        cherrypy.engine.start()
```
I found this _fix_ on SO [restart cherrypy](https://stackoverflow.com/questions/7254845/change-cherrypy-port-and-restart-web-server) and as far as I remember it helped me stop getting errors when trying to restart the server.

File: src/pybind/mgr/mgr_util.py
Comment (adk3798) [inline src/pybind/mgr/mgr_util.py:420]:
```suggestion
def check_port_in_use(addr: str, port: int) -> bool:
    """Checks if the port is available
    :raises PortAlreadyInUse: in case port is already in use
    :raises Exception: any generic error other than port already in use
    :return False if port is available
    """
```
NIT on naming and comment, feel free to take/leave whatever parts you agree/disagree with

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py:24]: I think I enabled it during the testing. We had some discussions to make it configurable so we can enable it in case we need to. By now I'll keep it disabeld.

File: src/pybind/mgr/cephadm/agent.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/agent.py:86]: Fixed

File: src/pybind/mgr/cephadm/ssl_cert_utils.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/ssl_cert_utils.py:14]: both of them are used by the `def verify_tls(crt, key)` which I moved from` mgr_util.py` to the new `ssl_cert_utils.py` (I think this function makes more sense in the new location rather than the mgr util).

File: src/pybind/mgr/mgr_util.py
Comment (rkachach) [reply src/pybind/mgr/mgr_util.py:420]: I removed the `return bool` because right now we rely on the `PortAlreadyInUse` exception to test if port is available or not.

File: src/pybind/mgr/cephadm/service_discovery.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/service_discovery.py:97]: technically the defaults are valid because the server will listen in `'0.0.0.0'` and a random port (if you pass 0) .. that what would happen if somebody tries to use the server without providing port/addr.

File: src/pybind/mgr/cephadm/http_server.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/http_server.py:48]: yeah, since this is the ctr I'm using the attribute directly (instead of the property) to avoid calling the underlying code. Outside the ctr the property is used instead.

File: src/pybind/mgr/cephadm/ssl_cert_utils.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/ssl_cert_utils.py:160]: yeah, that's the main concern is that we want this to be usable by all the modules. The issue with putting it here is that we end up in one of two scenarios: 1) we now just have duplicate code, both in places cephadm could access if it wanted, which doesn't seem correct or 2) we remove the duplicate in the mgr_utils in which case only cephadm can use this. Neither of them feel very good, so I'd prefer these functions just stay somewhere where all the modules can access them. If you want to move them around a bit and have them at the mgr level but in their own file it makes sense, but you then have to worry about changing every module that currently imports them from mgr_utils to now import from the new location. If you're willing to take that up then I'd be fine with you doing that but otherwise I think we should just leave them where they are.

File: src/pybind/mgr/cephadm/ssl_cert_utils.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/ssl_cert_utils.py:14]: okay, as said in the other thread, I don't really think these functions should be within cephadm and should stay at the mgr level, but I guess whether or not we import these libraries here is more dependent on that discussion than being its own issue

**Review by adk3798 — COMMENTED** [review]
LGTM, pending the discussion on where the `verify_xxx` util functions ought to live

File: src/pybind/mgr/cephadm/ssl_cert_utils.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/ssl_cert_utils.py:160]: @adk3798 yeah, my final goal is to have all these functions in their own file (at the mgr level) and have all the modules use this new location. As I don't want to mess other modules on this PR I'll remove them from the new file (`ssl_cert_utils.py`). In a separate PR (once this PR is merged) I'll move everything to the new `ssl_cert_utils`, move this file to the main mgr module and update all the modules using it.

File: src/pybind/mgr/mgr_util.py
Comment (rkachach) [reply src/pybind/mgr/mgr_util.py:420]: fixed

**Comment by rkachach** [comment]
jenkins test make check

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-09-21_01:18:31-orch:cephadm-wip-adk-testing-2022-09-20-1525-distro-default-smithi/

bunch of tests using the agent failed with timeouts. Did some local testing with this patch and saw issues with it locally as well. From the agent logs
```
Sep 21 13:03:09 vm-00 systemd[1]: Starting cephadm agent for cluster adc6bba6-39ab-11ed-accb-52540070db9f...
Sep 21 13:03:09 vm-00 systemd[1]: Started cephadm agent for cluster adc6bba6-39ab-11ed-accb-52540070db9f.
Sep 21 13:03:10 vm-00 bash[39060]: Verifying port 4721 ...
Sep 21 13:03:10 vm-00 bash[39060]: No cached ls output. Running full daemon ls
Sep 21 13:03:10 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:03:10 vm-00 bash[39060]: Inferring config /var/lib/ceph/adc6bba6-39ab-11ed-accb-52540070db9f/mon.vm-00/config
Sep 21 13:03:11 vm-00 bash[39060]: Using ceph image with id '28995033ff6e' and tag 'testing' created on 2022-09-21 12:47:05 +0000 UTC
Sep 21 13:03:11 vm-00 bash[39060]: quay.io/adk3798/ceph@sha256:5f9d25913c7d1dbda1037256220be054d0f681820cc9f98cd49a485a2478f4df
Sep 21 13:03:13 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:03:15 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:03:30 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:03:50 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:04:11 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:04:31 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
Sep 21 13:04:51 vm-00 bash[39060]: Failed to send metadata to mgr: HTTP Error 404: Not Found
```
what cephadm tried to connect to is `https://<target-ip>:<target-port>/data/` You can see what ip and port are configured in the agents config
```
[root@vm-00 ~]# cat /var/lib/ceph/adc6bba6-39ab-11ed-accb-52540070db9f/agent.vm-00/agent.json 
{"target_ip": "192.168.122.23", "target_port": 7150, "refresh_period": 20, "listener_port": 4721, "host": "vm-00", "device_enhanced_scan": "False"}
```
Trying to access the URL it would be building did give me a 404
```
[root@vm-00 ~]# curl --insecure https://192.168.122.23:7150/data
<!DOCTYPE html PUBLIC
"-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"></meta>
    <title>404 Not Found</title>
    <style type="text/css">
    #powered_by {
        margin-top: 20px;
        border-top: 2px solid black;
        font-style: italic;
    }

    #traceback {
        color: red;
    }
    </style>
</head>
    <body>
        <h2>404 Not Found</h2>
        <p>The path '/data' was not found.</p>
        <pre id="traceback"></pre>
    <div id="powered_by">
      <span>
        Powered by <a href="http://www.cherrypy.org">CherryPy 18.4.0</a>
      </span>
    </div>
    </body>
</html>
```

**Comment by rkachach** [comment]
> 
I think the issue probably has to do with having (or not) the `/` character at the end of the url `https://<target-ip>:<target-port>/data/`. I'll check and see if that's the issue.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-10-05_13:02:39-orch:cephadm-wip-adk-testing-2022-10-05-0715-distro-default-smithi/

2 failures:

- 1 seemingly infra thing installing an selinux package https://tracker.ceph.com/issues/57771
- Failure in initial part of staggered upgrade while manually redeploying the mgr daemons with new images. This will need more investigation but I don't think any of the PRs in the run were necessarily causing the issue.

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:2435]: found something interesting while testing this. Added a log statement to print out the old and new deps during `_check_daemons` for each daemon to see this working. For changing the service discovery port, it all worked as intended.
```
ceph config set mgr mgr/cephadm/service_discovery_port 8494
```
caused
```
2022-10-06T01:17:06.057476+0000 mgr.vm-00.rksyiw (mgr.34352) 854 : cephadm [INF] Changing service discovery port from 7159 to 8492...
2022-10-06T01:17:06.057618+0000 mgr.vm-00.rksyiw (mgr.34352) 855 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE Bus STOPPING
2022-10-06T01:17:06.076262+0000 mgr.vm-00.rksyiw (mgr.34352) 856 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('192.168.122.54', 7159)) shut down
2022-10-06T01:17:06.094348+0000 mgr.vm-00.rksyiw (mgr.34352) 857 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('192.168.122.54', 7150)) shut down
2022-10-06T01:17:06.094791+0000 mgr.vm-00.rksyiw (mgr.34352) 858 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE Bus STOPPED
2022-10-06T01:17:07.374859+0000 mgr.vm-00.rksyiw (mgr.34352) 859 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Bus STARTING
2022-10-06T01:17:07.489455+0000 mgr.vm-00.rksyiw (mgr.34352) 861 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Serving on https://192.168.122.54:7150
2022-10-06T01:17:07.618191+0000 mgr.vm-00.rksyiw (mgr.34352) 862 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Serving on https://192.168.122.54:8492
2022-10-06T01:17:07.619792+0000 mgr.vm-00.rksyiw (mgr.34352) 863 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Bus STARTED
```
and
```
2022-10-06T01:17:07.783871+0000 mgr.vm-00.rksyiw (mgr.34352) 872 : cephadm [INF] Reconfiguring prometheus.vm-01 (dependencies changed)...
2022-10-06T01:17:07.787574+0000 mgr.vm-00.rksyiw (mgr.34352) 873 : cephadm [INF] Reconfiguring daemon prometheus.vm-01 on vm-01
```
which is great, working as intended. However, I noticed in my new added log statement for the prometheus deps
```
2022-10-06T01:17:07.783808+0000 mgr.vm-00.rksyiw (mgr.34352) 871 : cephadm [ERR] daemon: prometheus.vm-01 | new deps: ['8492', '9283', 'alertmanager', 'mgr.vm-00.rksyiw', 'node-exporter'] | old deps: ['7159', '9283', 'alertmanager', 'mgr.vm-00.rksyiw', 'node-exporter']
```
as you can see, some of the deps are just `alertmanager` and `node-exporter`. Where as you look at the same log statement for grafana
```
2022-10-06T01:18:16.168248+0000 mgr.vm-00.rksyiw (mgr.34352) 928 : cephadm [ERR] daemon: grafana.vm-00 | new deps: ['prometheus.vm-01'] | old deps: ['prometheus.vm-01']
```
and its dep is the actual full prometheus daemon name, as it should be. I think https://github.com/ceph/ceph/commit/a0fe3ffdaaddd661f115de78ddae89b07b98a9ae introduced a regression where the prometheus deps are just getting the service type ("alertmanager", "node-exporter" or "ingress") as deps instead of the actual daemons in the service. Some further testing confirmed that we don't actually reconfig prometheus anymore if node-exporter daemons are moved to a different set of hosts.

Technically, this regression isn't caused by this PR, so I thought I'd leave it up to you if you want to fix it in this PR or in a follow up. If you want it as a follow up we can probably merge this as I think everything else is working fine. Let me know what you think.

**Review by adk3798 — APPROVED** [review]

File: src/pybind/mgr/cephadm/module.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/module.py:2435]: Thank you very much for the detailed testing and for the feedback. The goal from the PR https://github.com/ceph/ceph/pull/46400 was basically to make the whole prometheus configuration dynamic. That means we don't depend on a _static_ configuration anymore, instead, we just get the prometheus config (scrape and alert configs) dynamically by using the new service-discovery mechanism. So right now, our generated `prometheus.yml` looks like:

```
.....
scrape_configs:
  - job_name: 'ceph'
    honor_labels: true
    http_sd_configs:
    - url: https://192.168.100.100:8765/sd/prometheus/sd-config?service=mgr-prometheus
      tls_config:
        ca_file: root_cert.pem

  - job_name: 'node'
    http_sd_configs:
    - url: https://192.168.100.100:8765/sd/prometheus/sd-config?service=node-exporter
      tls_config:
        ca_file: root_cert.pem
```
So for example for node-exporter prometheus get the nodes list from the url `https://192.168.100.100:8765/sd/prometheus/sd-config?service=node-exporter`. So, if we curl this url we get:

```
[
  {
    "targets": [
      "192.168.100.100:9100"
    ],
    "labels": {
      "instance": "ceph-node-0"
    }
  },
  {
    "targets": [
      "192.168.100.101:9100"
    ],
    "labels": {
      "instance": "ceph-node-1"
    }
  },
  {
    "targets": [
      "192.168.100.102:9100"
    ],
    "labels": {
      "instance": "ceph-node-2"
    }
  }
]

```
The thing is, I think with this new dynamic configuration we don't have to adjust the promteheus deps to depend on the daemons names anymore (maybe we can also remove the service-type! as it's in fact useless) as we just get the conf dynamically. If you move some node-exporter daemon to some other host we should just get the new list from the above url updated and that's whay prometheus will be using. 

WDYT?

PD: Regardless of whether we need an additional PR to fix this dep issues or not I also think it's better idea to merge this PR and then open a new one to address any deps related issues.

Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:2435]: okay, in that case we should definitely just remove the deps. Having the service names in there isn't very useful. I'm okay with this being done in a separate PR

**Pull request merged**
